### ISSUE TITLE
music21's import conventions prevent using PyInstaller

### ISSUE THREAD

**Issue by keeganNull** [cuthbertLab/music21#881]
I have a program that uses the music21. When I execute the program of .py , it's OK. But when I use the pyinstaller to pack the program and execute it. It shows that 
![image](https://user-images.githubusercontent.com/69446006/109382107-5741b780-7919-11eb-848f-cdff0038085c.png)
and the code:
![image](https://user-images.githubusercontent.com/69446006/109382123-6d4f7800-7919-11eb-97a8-28669fe37d80.png)

**Comment by keeganNull** [comment]
And I just try a program like this.
![image](https://user-images.githubusercontent.com/69446006/109382446-74778580-791b-11eb-9aa2-1c13e491c1f7.png)
still like this:
![image](https://user-images.githubusercontent.com/69446006/109382455-7fcab100-791b-11eb-9b6a-5c514b7e12f5.png)

**Comment by jacobtylerwalls** [comment]
Hi -- in the future, please use the provided issue template. For instance, what version of music21 are you using? There was a fix for `importlib` issues in 6.7.0.

**Comment by keeganNull** [comment]
> Hi -- in the future, please use the provided issue template. For instance, what version of music21 are you using? There was a fix for `importlib` issues in 6.7.0.
Sorry for I didn't use the template.
My music21 version is 6.7.1. python version in 3.7.4. 
As for I'm not very good at python,  I don't understand what are you talking about. Should I update the 'importlib' or change the way I import 'music21'?

**Comment by jacobtylerwalls** [comment]
Try this:

'pip install numpy==1.19.3'

Source:


https://stackoverflow.com/questions/64772965/pyinstaller-executable-file-throws-an-error-when-opening-on-another-pc

**Comment by keeganNull** [comment]
> Try this:
> 
> 'pip install numpy==1.19.3'
> 
> Source:
> 
> https://stackoverflow.com/questions/64772965/pyinstaller-executable-file-throws-an-error-when-opening-on-another-pc

Sorry. I tried it. But it didn't work.  According to your source, it's the problem when the program executes on different PCs.But my error may is caused by the 'music21'.
And the error : module 'music21' has no attribute 'alpha'.

**Comment by mscuthbert** [comment]
I don't know what pyinstaller is -- we recommend using `pip` which is the default Python installer.

**Comment by keeganNull** [comment]
> I don't know what pyinstaller is -- we recommend using `pip` which is the default Python installer.

PyInstaller freezes (packages) Python applications into stand-alone executables.

**Comment by keeganNull** [comment]
I just want to report a bug. 
Do as this:
 1. pip install Pyinstaller
 2. create a .py named test.py with a single line code :' import music21'
 3. use ' pyinstaller -F test.py' to freeze the test.py
 4. go to the file named 'dist' and execute the test.exe
 
and then you will see the error. And the error happens on two different PC.

**Comment by jacobtylerwalls** [comment]
Thanks -- Michael and I are expressing the view that we don't know if it's an issue with PyInstaller or music21 until someone volunteers to do more investigation. PyInstaller's page enumerates the packages it works with; ours isn't on the list. It's nice that NumPy and Django play well with it, but music21 just doesn't have the audience they do yet.

It's not as simple as "hey, the name alpha is missing somewhere", it's a circular import issue given however PyInstaller freezes imports:

``` 
File "PyInstaller/loader/pyimod03_importers.py", line 531, in exec_module
  File "music21/__init__.py", line 195, in <module>
AttributeError: partially initialized module 'music21' has no attribute 'alpha' (most likely due to a circular import)
```

Your steps to reproduce were very helpful -- and so after 10 or so minutes I could produce no errors by commenting out this line:
https://github.com/cuthbertLab/music21/blob/ef4bdd6ae787ffb189486a126e6f1a2ee0b556c7/music21/__init__.py#L195

You could give this a try and let us know if it works with PyInstaller, but at the most I think it would just advance the investigation -- I don't think it would be a fix, because import statements across the entire codebase would have to be retooled. I'm not volunteering for that at this time, but I'm happy to hear the result of further investigation by anyone.

**Comment by jacobtylerwalls** [comment]
@mscuthbert Do you recall why that wildcard import is there? `from music21 import *` still works fine in a shell, and the unit tests pass, it's just the doctests that fail with NameErrors if I comment out that wildcard import.

This is the diff to get the editorial module doctests to run:

```diff --git a/music21/editorial.py b/music21/editorial.py
index 64f6e50e1..ccccb7178 100644
--- a/music21/editorial.py
+++ b/music21/editorial.py
@@ -20,6 +20,7 @@ pre-defined aspects might be removed from documentation in the future.
 
 Access an editorial object by calling `.editorial` on any music21 object:
 
+>>> from music21 import clef
 >>> c = clef.TrebleClef()
 >>> ed = c.editorial
 >>> ed
@@ -29,6 +30,7 @@ The object is lazily created on first access.
 To see if there is any existing editorial information without creating
 the object, call `.hasEditorialInformation`
 
+>>> from music21 import note
 >>> n = note.Note('C#4')
 >>> n.hasEditorialInformation
 False
@@ -75,6 +77,7 @@ class Editorial(prebase.ProtoM21Object, dict):
     the editorial suggestion to sing F-sharp as a "musica ficta" accidental
     object:
 
+    >>> from music21 import note, pitch
     >>> fictaSharp = pitch.Accidental('sharp')
     >>> n = note.Note('F')
     >>> n.editorial.ficta = fictaSharp
@@ -143,6 +146,7 @@ class Comment(prebase.ProtoM21Object, style.StyleMixin):
     >>> c.isFootnote = True
     >>> c.levelInformation = 'musicological'
 
+    >>> from music21 import note
     >>> n = note.Note('C#4')
     >>> n.editorial.footnotes.append(c)
     >>> n.editorial.footnotes[0]      
```

Putting the imports at the top of the module didn't work because there were circular imports with music21.base -- but then again, if the hood is ever popped on this, maybe it's faster to refactor music21.base than to put these imports on every single doctest.

**Comment by mscuthbert** [comment]
I can't remember why it's there, but it used to be necessary.  It might not be -- I haven't tested trying without it for a long time.  But I'm definitely not going to change the doctest style :-)  -- the doc tests follow the recommendation for users to run "from music21 import *" once and then access music21 modules with their names without music21.

**Comment by mscuthbert** [comment]
In any case -- the current solution to the bug is "do not use PyInstaller -- not supported for music21.  If you must use it, please purchase a consulting contract."

**Comment by keeganNull** [comment]
I tried `from music21 import * `, it still didn't work. And yes I have to use your music21, it's a nice package. That's ok, I can wait.

**Comment by jacobtylerwalls** [comment]
Hi @keeganNull this appears to be fixed by 91e677d8189cb5164b7a7db0eb3b7a8e64dee9b7 (thanks Myke!) so I'm closing this issue. If you need this before 7.3 is released, you can use pip to install from the latest commit on master using the project's URL.
